x/dex/module.go

```diff
@@ -246,25 +245,93 @@ func (am AppModule) beginBlockForContract(ctx sdk.Context, contractAddr string)
 // EndBlock executes all ABCI EndBlock logic respective to the capability module. It
 // returns no validator updates.
 func (am AppModule) EndBlock(ctx sdk.Context, _ abci.RequestEndBlock) []abci.ValidatorUpdate {
-	for _, contractAddr := range am.getAllContractAddresses(ctx) {
-		ctx.Logger().Info(fmt.Sprintf("End block for %s", contractAddr))
-		am.endBlockForContract(ctx, contractAddr)
+	validContractAddresses := map[string]types.ContractInfo{}
+	for _, contractInfo := range am.getAllContractInfo(ctx) {
+		validContractAddresses[contractInfo.ContractAddr] = contractInfo
+	}
+	// Each iteration is atomic. If an iteration finishes without any error, it will return,
+	// otherwise it will rollback any state change, filter out contracts that cause the error,
+	// and proceed to the next iteration. The loop is guaranteed to finish since
+	// `validContractAddresses` will always decrease in size every iteration.
+	for len(validContractAddresses) > 0 {
```

Comment:
@philipsu522 so if it fails for *some contracts*, we will delete those contracts from `validContractAddress` and rollback the store (not really rollback but more like discard the corrupt cached store and use a fresh one), and rerun the logics for contracts that are not failing. If there are no inter-dependencies among contracts, this loop would iterate at most twice (the first time it would weed out all contracts that are failing, and the second time it would succeed for all remaining). But if there is any inter-dependency (which we don't have a way to prevent), some contracts that succeeded in the first iter might fail in the second iter, so we keep the loop going until either all remaining succeed or everyone fails.

Reply:
Gotcha. I'm just worried some edge case happens that we don't catch where this loop ends up going forever. A cleaner way might be to explicitly recall the method with a flag that's
```
second_call_remove_invalid=False
```
and when it's the second time, pass it true. That way we guarantee it's only called twice. I"ll leave that up to you thought!

Reply:
oh i meant that it's possible for have inter-dependency between contracts and thus possible for it to run more than two times. But i get what you are saying and I can probably add a counter such that it runs at most len(registeredContracts) times to guarantee termination